igvm/hypervisor.py

```diff
@@ -555,15 +556,77 @@ def vm_block_device_name(self):
         the guest OS"""
         return 'vda1'
 
-    def migrate_vm(
-        self, vm, target_hypervisor, offline, offline_transport, transaction,
-        no_shutdown,
+    def check_migrate_vm(
+        self, vm: VM, offline: bool, offline_transport: str,
+        disk_size: int = 0,
     ):
-        if offline_transport not in ['netcat', 'drbd']:
+        if offline_transport not in ['netcat', 'drbd', 'xfs']:
             raise StorageError(
                 'Unknown offline transport method {}!'
                 .format(offline_transport)
             )
+        disk_size = disk_size if 0 < disk_size else 0
+        if not (offline and offline_transport == 'xfs') and 0 < disk_size:
+            raise StorageError(
+                'disk_size can be applied only with offline transport xfs!'
+            )
+        allocated_space = vm.dataset_obj['disk_size_gib'] - vm.disk_free()
+        if 0 < disk_size < allocated_space:
+            raise StorageError(
+                'disk_size is lower than allocated space: {} < {}!'
+                .format(disk_size, allocated_space)
+            )
+
+    def migrate_disk_size(
+        self, vm: VM, offline: bool, offline_transport: str,
+        disk_size: int = 0,
+    ) -> int:
+        self.check_migrate_vm(vm, offline, offline_transport, disk_size)
+        disk_size = disk_size if 0 < disk_size else 0
+        return disk_size or vm.dataset_obj['disk_size_gib']
+
+    def migrate_change_disk_size(
+        self, vm: VM, offline: bool, offline_transport: str,
+        transaction, disk_size: int = 0,
+    ):
+        size = self.migrate_disk_size(vm, offline,
+                                      offline_transport, disk_size)
+        if size == vm.dataset_obj['disk_size_gib']:
+            return
+
+        old_size = vm.dataset_obj['disk_size_gib']
+        vm.dataset_obj['disk_size_gib'] = size
+        vm.dataset_obj.commit()
+
+        if transaction:
+            def restore_size():
+                vm.dataset_obj['disk_size_gib'] = old_size
+                vm.dataset_obj.commit()
+            transaction.on_rollback('reset_disk_size', restore_size)
+
+
+    def migrate_wait_shutdown(
+        self, vm: VM, no_shutdown: bool, transaction: Transaction,
+    ):
+        vm.set_state('maintenance', transaction=transaction)
+        if vm.is_running():
+            if no_shutdown:
+                log.info('Please shut down the VM manually now')
+                vm.wait_for_running(running=False, timeout=86400)
```

Comment:
I don't think it is realistic that we have to wait 24 hours to shutdown a VM. I suggest using a smaller timeout such as 5 minutes and check again if the VM was shut down and if not raise an Exception.

Reply:
This method is an unchanged snippet from [1](https://github.com/innogames/igvm/pull/176/commits/5fff38a34b71643951dc72192c65625535f06240#diff-d587443efd90ee0cdc812d478981451c4b4a15000f584b1fc06ea3256d154fddL609-L619) and [2](https://github.com/innogames/igvm/pull/176/commits/5fff38a34b71643951dc72192c65625535f06240#diff-d587443efd90ee0cdc812d478981451c4b4a15000f584b1fc06ea3256d154fddL622-L632)

Reply:
This function was designed for migration of Graphite hosts and they needed a very long time to shut down. Let's just keep it as it is.

Reply:
Okay but lets at least tell the user than how long we will wait.

Reply:
Added to argparse help

Reply:
Fair enough.